File: serial-monitor/src/main/java/com/intellij/plugins/serialmonitor/ui/actions/EditSettingsAction.java
package com.intellij.plugins.serialmonitor.ui.actions;

import com.intellij.icons.AllIcons;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.DumbAwareAction;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.plugins.serialmonitor.ui.SerialProfileConfigurable;
import com.intellij.plugins.serialmonitor.ui.console.SerialConnectable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

import static com.intellij.plugins.serialmonitor.ui.SerialMonitorBundle.message;

/**
 * @author Dmitry_Cherkas
 */
public class EditSettingsAction extends DumbAwareAction {

  private final @NotNull String myName;
  private final SerialConnectable<?> myConnectable;

  public EditSettingsAction(@NotNull String name, SerialConnectable<?> connectable) {
    super(message("edit-settings.title"), message("edit-settings.tooltip"), AllIcons.General.Settings);
    myName = name;
    myConnectable = connectable;
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent e) {

    boolean okClicked = new SettingsDialog(e).showAndGet();
    if (okClicked) {
      myConnectable.reconnect();
    }
  }

  private class SettingsDialog extends DialogWrapper {
    private SettingsDialog(AnActionEvent e) {
      super(e.getProject());
      setTitle(message("dialog.title.serial.port.settings", myName));
      init();
    }

    @Override
    protected @Nullable JComponent createCenterPanel() {
      return SerialProfileConfigurable.Companion.createSettingsPanel(myConnectable.getPortProfile(), false, false);
    }
  }
}

File: serial-monitor/src/main/java/com/intellij/plugins/serialmonitor/ui/actions/EditSettingsAction.kt
package com.intellij.plugins.serialmonitor.ui.actions

import com.intellij.icons.AllIcons
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.DumbAwareAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.plugins.serialmonitor.ui.SerialMonitor
import com.intellij.plugins.serialmonitor.ui.SerialMonitorBundle
import com.intellij.plugins.serialmonitor.ui.serialSettings
import com.intellij.ui.JBColor
import com.intellij.ui.dsl.builder.Align
import com.intellij.ui.dsl.builder.panel
import org.jetbrains.annotations.Nls
import javax.swing.BorderFactory
import javax.swing.JComponent

/**
 * @author Dmitry_Cherkas, Ilia Motornyi
 */
internal class EditSettingsAction(private val myName: @Nls String, private val serialMonitor: SerialMonitor) :
  DumbAwareAction(SerialMonitorBundle.message("edit-settings.title"),
                  SerialMonitorBundle.message("edit-settings.tooltip"),
                  AllIcons.General.Settings) {
  override fun actionPerformed(e: AnActionEvent) {

    val settingsDialog = SettingsDialog(e.project)
    val okClicked: Boolean = settingsDialog.showAndGet()
    if (okClicked) {
      serialMonitor.notifyProfileChanged()
    }
  }

  private inner class SettingsDialog(project: Project?) : DialogWrapper(project, false, IdeModalityType.IDE/*todo change to PROJECT when platform is fixed*/) {
    init {
      title = SerialMonitorBundle.message("dialog.title.serial.port.settings", myName)
      init()
    }

    override fun createCenterPanel(): JComponent {
      return panel {
        row {
          label(serialMonitor.portProfile.portName).align(Align.FILL)
            .resizableColumn()
            .label(SerialMonitorBundle.message("label.port"))
            .applyToComponent {
              border = BorderFactory.createLineBorder(JBColor.border())
            }
        }.layout(com.intellij.ui.dsl.builder.RowLayout.LABEL_ALIGNED)
        serialSettings(disposable, serialMonitor.portProfile) { }
      }
    }
  }
}
